pkg/network/node/networkpolicy.go

```diff
@@ -276,15 +281,32 @@ func (np *networkPolicyPlugin) DeleteNetNamespace(netns *osdnv1.NetNamespace) {
 	np.updateMatchCache(npns)
 }
 
-func (np *networkPolicyPlugin) SetUpPod(podIP string) error {
-	otx := np.node.oc.NewTransaction()
-	// when a pod is created isolate the pod until the current network policy rules can be
-	// evaluated
-	otx.AddFlow("table=27, priority=500, cookie=1, ip, nw_src=%s, actions=drop", podIP)
-	otx.AddFlow("table=80, priority=500, cookie=1, ip, nw_src=%s, actions=drop", podIP)
-	if err := otx.Commit(); err != nil {
-		return fmt.Errorf("Error syncing OVS flows to isolate pods %v", err)
+// SetUpPod only exists to ensure network policy is applied to a newly created pod before CNI ADD sends a response.
+// Network Policies are usually updated asynchronously, but for pod creation we need to wait for network policy to
+// be properly applied. Therefore, this function is a synchronous version of refreshPodNetworkPolicies, that calls
+// np.syncFlows() directly and not via np.runner.Run()
+func (np *networkPolicyPlugin) SetUpPod(pod *corev1.Pod, podIP string) error {
+	np.lock.Lock()
+
+	np.localPodIPs[string(pod.UID)] = podIP
+	podNs := np.namespacesByName[pod.Namespace]
+
+	// this code is a combination of refreshPodNetworkPolicies and syncNamespace
+	for _, npns := range np.namespaces {
+		for _, npp := range npns.policies {
+			if (npp.watchesOwnPods && npns == podNs) || npp.watchesAllPods {
+				npns.mustRecalculate = true
+			}
+		}
+		if npns.mustRecalculate && npns.inUse {
+			if !npns.mustSync {
+				npns.mustSync = true
+			}
+		}
 	}
+	np.lock.Unlock()
+	np.syncFlows()
```

Comment:
I am not sure I get this, waiting for "pod creation to have to block on that" is exactly what I want - I want to update network policies first and only then reply to `CNI ADD` that's why this fix is required at all, right?

Reply:
Not exactly. The current timeline is:

1. kubelet creates the pod
2. kubelet calls CNI ADD; openshift-sdn assigns IP, attaches pod to network with `action=drop` flows, returns from CNI ADD
3. kubelet runs `postStart` hook, waits for it to complete
4. kubelet sets `pod.Status.PodIP`
5. openshift-sdn programs correct NetworkPolicy, removes `action=drop` flows

Step 5 needs to happen before step 4, but it doesn't have to happen during step 2; it can happen in parallel with step 3. (The `postStart` hook needs to be able to deal with some initial network flakiness anyway, since there's no guarantee that _all_ nodes have updated before it starts.)

Reply:
I think CNI protocol implies that when CNI ADD is handled, pod should have network connectivity https://github.com/containernetworking/cni/blob/main/SPEC.md#overview-1
Also, I don't see what is the point in trying to send CNI ADD reply earlier and don't block pod creation, if pod still won't have connectivity, even if will be "created" earlier?

Reply:
I don't think this parallelization will work:
`SetUpPod` and `syncFlows` both trying to take `np.lock`, and since sdn handles add pod requests sequentially, `syncFlows` will be called before the next `SetUpPod` (verified on a test cluster), and lock queue is FIFO. So to really use only 1 call of `syncFlows` we need to call all `SetUpPod`s for e.g. 5 pods before `syncFlows` is called, which doesn't seem possible.
Am I missing something?